python_scripts/convert_to_torch.py

```diff
@@ -0,0 +1,162 @@
+    out_channels = decode_single_int32(f)
+    bias_data = torch.Tensor(decode_float32(f, out_channels))
+    kernels = [decode_matrix(f) for _ in range(in_channels * out_channels)]
+    kernel_size = (kernels[0].size(0), kernels[0].size(1))
+    conv_layer = Conv2d(in_channels=in_channels,
+                        out_channels=out_channels,
+                        kernel_size=kernel_size)
+
+    # Initialize weight
+    weight_data = torch.cat(kernels).view([in_channels, out_channels] + list(kernel_size))
+    weight_data = weight_data.permute(1, 0, 2, 3)
+    assert conv_layer.weight.shape == weight_data.shape
+    assert conv_layer.bias.shape == bias_data.shape
+    conv_layer.weight.data = weight_data
+    conv_layer.bias.data = bias_data
+    return conv_layer
+
+
+def decode_max_pooling(f):
+    """ Return MaxPool2D """
+    kernel_x = decode_single_int32(f)
+    kernel_y = decode_single_int32(f)
+    stride_x = decode_single_int32(f)
+    stride_y = decode_single_int32(f)
+    return MaxPool2d(kernel_size=[kernel_x, kernel_y],
+                     stride=[stride_x, stride_y])
+
+
+def decode_linear_layer(f):
+    """ Return a linear layer """
+    bias_data = decode_matrix(f)
+    bias_data = bias_data.squeeze(-1)
+    weight_data = decode_matrix(f)
+    linear_layer = LinearChannelWise(in_features=weight_data.size(0),
+                                     out_features=weight_data.size(1))
+    weight_data = weight_data.permute(1, 0)
+    assert linear_layer.weight.data.shape == weight_data.shape
+    assert linear_layer.bias.data.shape == bias_data.shape
+    linear_layer.weight.data = weight_data
+    linear_layer.bias.data = bias_data
+    return linear_layer
+
+
+def decode_prelu(f):
+    """ Return a PReLU """
+    weight_data = decode_matrix(f)
+    weight_data = weight_data.squeeze(-1)
+    prelu = PReLU(num_parameters=weight_data.shape.numel())
+    assert prelu.weight.data.shape == weight_data.shape
+    prelu.weight.data = weight_data
+    return prelu
+
+
+def decode_cnn(f):
+    """ Return a torch.Module """
+    cnn = torch.nn.Sequential()
+    depths = decode_single_int32(f)
+    print('Depth: {}'.format(depths))
+    for layer_idx in range(depths):
+        layer_type = decode_single_int32(f)
+        if layer_type == 0:
+            layer = decode_conv_layer(f)
+        elif layer_type == 1:
+            layer = decode_max_pooling(f)
+        elif layer_type == 2:
+            layer = decode_linear_layer(f)
+        elif layer_type == 3:
+            layer = decode_prelu(f)
+        else:
+            raise ValueError('Invalid layer type')
+        cnn.add_module('layer_{}'.format(layer_idx), layer)
+    return cnn
+
+if __name__ == '__main__':
+    pnet_path = '/home/yuchen/QuindiTech/OpenFace/lib/local/LandmarkDetector/model/mtcnn_detector/PNet.dat'
+    with open(pnet_path, 'rb') as f:
+        pnet = decode_cnn(f)
+    print(pnet)
+    imgs = torch.rand(1, 3, 128, 128)
+    result = pnet(imgs)
```

Comment:
Have you confirmed that this results in the same output as the Matlab or C++ versions?